File: backend/app/schemas/circle.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from .user import UserResponse

class CircleBase(BaseModel):
    name: str = Field(..., max_length=100)
    description: Optional[str] = None

class CircleCreate(CircleBase):
    pass

class CircleUpdate(CircleBase):
    name: Optional[str] = Field(None, max_length=100)

class Circle(CircleBase):
    id: int
    invite_code: str
    creator_id: int
    is_active: bool
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

class CircleResponse(BaseModel):
    id: int
    name: str
    description: Optional[str]
    invite_code: str
    creator_id: int
    is_active: bool
    created_at: datetime
    member_count: Optional[int] = 0

    class Config:
        from_attributes = True

class CircleMemberBase(BaseModel):
    role: str = Field(default="member", pattern="^(admin|member)$")

class CircleMemberCreate(CircleMemberBase):
    circle_id: int
    user_id: int

class CircleMember(CircleMemberBase):
    id: int
    circle_id: int
    user_id: int
    joined_at: datetime
    user: Optional[UserResponse] = None

    class Config:
        from_attributes = True

class CircleJoinRequest(BaseModel):
    invite_code: str = Field(..., max_length=20)

File: circly-app/src/types/circle.ts
import { UserResponse } from './user';

// Circle types
export interface Circle {
  id: number;
  name: string;
  description?: string;
  invite_code: string;
  creator_id: number;
  is_active: boolean;
  created_at: string;
  updated_at?: string;
}

export interface CircleResponse {
  id: number;
  name: string;
  description?: string;
  invite_code: string;
  creator_id: number;
  is_active: boolean;
  created_at: string;
  member_count: number;
}

export interface CircleCreate {
  name: string;
  description?: string;
}

export interface CircleUpdate {
  name?: string;
  description?: string;
}

// Circle Member types
export interface CircleMember {
  id: number;
  circle_id: number;
  user_id: number;
  role: 'admin' | 'member';
  joined_at: string;
  user?: UserResponse;
}

export interface CircleMemberCreate {
  circle_id: number;
  user_id: number;
  role?: 'admin' | 'member';
}

export interface CircleJoinRequest {
  invite_code: string;
}

// Circle state
export interface CircleState {
  circles: CircleResponse[];
  currentCircle: CircleResponse | null;
  members: CircleMember[];
  loading: boolean;
  error: string | null;
}